### ISSUE TITLE
1st

### ISSUE THREAD

**Issue by DaleLeeCoding** [DaleLeeCoding/daily#1]
# Stable diffusion(High-Resolution Image Synthesis with Latent Diffusion Models)

[GitHub - CompVis/stable-diffusion: A latent text-to-image diffusion model](https://github.com/CompVis/stable-diffusion)](https://github.com/CompVis/stable-diffusion)

[CompVis/stable-diffusion · Hugging Face](https://huggingface.co/CompVis/stable-diffusion)](https://huggingface.co/CompVis/stable-diffusion)
![image](https://user-images.githubusercontent.com/69357689/193794454-29353439-b7c4-4863-943a-fd305bc71afa.png)

# Info

- The model was pretrained on 256x256 images and then finetuned on 512x512 images from a subset of the [[LAION-5B](https://laion.ai/blog/laion-5b/)](https://laion.ai/blog/laion-5b/) database.
  - ckpt: **`[stable-diffusion-v1-4](https://huggingface.co/CompVis/stable-diffusion-v1-4)`** (최신)
      1. 237k steps at resolution `256x256` on [[laion2B-en](https://huggingface.co/datasets/laion/laion2B-en)](https://huggingface.co/datasets/laion/laion2B-en). 
      2. 194k steps at resolution `512x512` on [[laion-high-resolution](https://huggingface.co/datasets/laion/laion-high-resolution)](https://huggingface.co/datasets/laion/laion-high-resolution) (170M examples from LAION-5B with resolution `>= 1024x1024`)
      3. 515k steps at resolution `512x512` on [[laion-aesthetics v2 5+](https://laion.ai/blog/laion-aesthetics/)](https://laion.ai/blog/laion-aesthetics/) (a subset of laion2B-en with estimated aesthetics score `> 5.0`, and additionally filtered to images with an original size `>= 512x512`, and an estimated watermark probability `< 0.5`. The watermark estimate is from the [[LAION-5B](https://laion.ai/blog/laion-5b/)](https://laion.ai/blog/laion-5b/) metadata, the aesthetics score is estimated using the [[LAION-Aesthetics Predictor V2](https://github.com/christophschuhmann/improved-aesthetic-predictor)](https://github.com/christophschuhmann/improved-aesthetic-predictor))
      4. 225k steps at resolution `512x512` on "laion-aesthetics v2 5+" and 10% dropping of the text-conditioning to improve [[classifier-free guidance sampling](https://arxiv.org/abs/2207.12598)](https://arxiv.org/abs/2207.12598)
- **frozen CLIP ViT-L/14 text encoder(123M)** to condition the model on text prompts
- **U-Net encoder(860M)**
- **Variational Autoencoder** → Open Images pretrained downsampling-factor 8 autoencoder

## 1차 목표)

**Text encoder : frozen CLIP ViT-L/14 → Korean text encoder like KELIP**

## Diffusion model fine-tuning

1. LN
2. top or bottom
3. Residual connection level Weight mix-up

### Upsampling

[https://github.com/openai/guided-diffusion](https://github.com/openai/guided-diffusion)

- 64x64 -> 256x256 upsampler: [[64_256_upsampler.pt](https://openaipublic.blob.core.windows.net/diffusion/jul-2021/64_256_upsampler.pt)](https://openaipublic.blob.core.windows.net/diffusion/jul-2021/64_256_upsampler.pt)
- 128x128 -> 512x512 upsampler: [[128_512_upsampler.pt](https://openaipublic.blob.core.windows.net/diffusion/jul-2021/128_512_upsampler.pt)](https://openaipublic.blob.core.windows.net/diffusion/jul-2021/128_512_upsampler.pt)

224x224 → 128x128 → 512x512 과정으로 feasibility 확인

# BEiT-3 training

[unilm/beit3 at master · microsoft/unilm](https://github.com/microsoft/unilm/tree/master/beit3)](https://github.com/microsoft/unilm/tree/master/beit3)

![image](https://user-images.githubusercontent.com/69357689/193794532-3a851ce0-fd50-431c-977c-490b59f465ee.png)

## Backbone : VLMo

- VLMo
  
![image](https://user-images.githubusercontent.com/69357689/193794798-5f1b1ccd-bb3c-4fc5-b141-0fbc1c6a7da9.png)
  
**MOME** Transformer introduces mixture of modality experts as a substitute of the feed forward network of standard Transformer.

**shared self-attention module** across different modalities to align the contents

**pool of modality experts** to capture modality-specific information

![image](https://user-images.githubusercontent.com/69357689/193794846-77922177-dd23-4090-9396-169b967e5755.png)

  
  ### Stagewise Pre-Training process
  
  1. pretrain the vision expert (V-FFN) and self-attention module on large-scale image-only data as in BEIT
  2. the parameters of vision expert and self-attention module are frozen, and we train  the language expert(L-FFN) by masked language modeling on large amounts of text-only data
  3. Finally, we train the whole model with vision-language pre-training
  - For images, the input resolution is 224 × 224 and the patch size is 16 × 16 during pre-training
  - RandAugment for IMG
  - whole word masking for MLM
  - Global Hard Negative Mining (ALBEF의 경우 single GPU에서만 mining)
  
  ### Dataset - 10M image-text pairs in the pre-training data and 4M images
  
  1. Conceptual Captions (CC)
  2. SBU Captions
  3. COCO
  4. Visual Genome (VG)
  
  Text-only pretraining data
  
  1. English Wikipedia
  2. BookCorpus
  
  ### Model Architecture
  
  same network configuration as ViT and BEIT.
  
  - VLMO-Base
      - 12-layer Transformer blocks with 768 hidden size and 12 attention heads
      - vision-language expert on the **top two** Transformer layers
      - two days using 64 Nvidia Tesla V100 32GB GPU cards
  - VLMO-Large
      - 24-layer Transformer network with 1024 hidden size and 16 attention heads
      - vision-language expert on the **top three** layers
      - three days using 128 Nvidia Tesla V100 32GB GPU cards
  
  [](https://arxiv.org/pdf/2111.02358.pdf)
  

## Pretraining : unified masked data modeling from VL-BEiT

- VL-BEiT
![image](https://user-images.githubusercontent.com/69357689/193794934-1991da5e-3647-44c9-ad95-cf93ec2844cd.png)
  
  one unified generative pretraining task 
  one shared bidirectional Transformer
  one-stage training from scratch
  
  [](https://arxiv.org/pdf/2206.01127.pdf)
  
- Text data is tokenized by a SentencePiece tokenizer
- Image data is tokenized by the tokenizer of BEIT v2
- Mask-then-predict task 가 image-text contrast, image-text matching, and word-patch/region alignment 등의 다른 pretraining task보다 더 적은 batch size를 사용한다는 이점을 가짐.

ViT-giant layout from *Scaling Vision Transformers*

model consists of a 40-layer Multiway Transformer with 1408 hidden size, 6144 intermediate size, and 16 attention heads. All layers contain both vision experts and language experts. Vision-language experts are also employed in the top three Multiway Transformer layers.

## Data  : 21M image-text pairs and 15M images
![image](https://user-images.githubusercontent.com/69357689/193795027-bbe01e8d-a4fd-4b74-8c92-f967bb0b2423.png)

# Landmark Recognition survey

[https://github.com/cvdfoundation/google-landmark](https://github.com/cvdfoundation/google-landmark)

- Dataset info(Google Landmarks Dataset v2 (GLDv2))
    - 5M, 200K classes, the number of training examples per class may not be very large
        - train(4M) + index(0.7M) + test(0.1M)
        - clean dataset(1.6M) + noisy dataset(3.2M)
    - Metric: Global Average Precision (GAP) at (k), where (k=1). This metric is also known as micro Average Precision
    - Metric for Recognition and Retrieval computation code - OK

# ****Google Landmark Recognition competition****

# ****Model Architectures****

**leverage both, global and local image information**

1. **end2end**
    1. combine local and global descriptor into a single image embedding
2. **two-stage**
    1. global (preliminary recognition)→ local (post reranking using spatial verification)

## ****DOLG: Single-Stage Image Retrieval with Deep Orthogonal Fusion of Local and Global Features - ICCV2021 - Baidu****

![image](https://user-images.githubusercontent.com/69357689/193795839-bd92c3c7-9859-4ff3-80ef-93c9ca1b0243.png)

train local and global descriptors also simultaneously but additionally fusing them within the model into a single descriptor

## Swin transformer - ICCV2021

![image](https://user-images.githubusercontent.com/69357689/193795862-b84f306e-43de-48cf-b554-160d589e7fa4.png)

- 작은 단위의 patch로부터 시작해서 점점 patch들을 merge해나가는 방식
- patch들을 합치며 계층적인 구조로 각 단계마다 representation을 갖기 때문에 다양한 크기의 entity, scale을 다루어야 하는 비전 분야에서 좋은 성능을 낸다.
- Shifted Window 개념을 통해 이미지 Size에 대해서 Linear하게 연산량이 증가하는 방식의 Backbone을 제안
- Window 내부에 존재하는 Patch끼리만 Self Attention 연산을 수행
    - Window Multi-Head Self Attention
        
        윈도우에 있는 패치들끼리만 self-attention 연산을 수행
        
    - Shifted Window Multi-Head Self Attention
        
        Window간의 Connection을 학습하면서, 연산량의 효율성은 그대로 유지
        
- ViT와 달리 Positional Embedding을 입력부분에서 추가하지 않고 Relative Position Bias를 활용
       

# solution

[https://github.com/ChristofHenkel/kaggle-landmark-2021-1st-place](https://github.com/ChristofHenkel/kaggle-landmark-2021-1st-place)

# Models

- DOLG + EfficientNet
- Hybrid-Swin-Transformers

Both especially strong architectures for integrating **local** and **global** descriptors of a landmark into a single feature vector in a single-stage manner

**DOLG - ICCV2021**

![image](https://user-images.githubusercontent.com/69357689/193795959-b863f757-5da7-4af7-a71c-cc84a6cf402e.png)

**Hybrid Swin-T: half a CNN encoder(EfficientNet) with a transformer ending**

![image](https://user-images.githubusercontent.com/69357689/193795986-f73fa9c2-8620-4972-9d3c-ccdae002cf70.png)

- transformers are like graph neural nets on image patches **connecting distant local features with each other.** → CNN에서는 부족한 부분이지만 landmark recognition/retrieval에서는 굉장히 중요

# Training

- Adam , cosine annealing
- augmentation
    
    ```python
    cfg.train_aug = A.Compose([
            A.HorizontalFlip(p=0.5),
            A.ImageCompression(quality_lower=99, quality_upper=100),
            A.ShiftScaleRotate(shift_limit=0.2, scale_limit=0.2, rotate_limit=10, border_mode=0, p=0.7),
            A.Resize(image_size, image_size),
            A.Cutout(max_h_size=int(image_size * 0.4), max_w_size=int(image_size * 0.4), num_holes=1, p=0.5),
        ])
    
    cfg.val_aug = A.Compose([A.Resize(image_size, image_size),])
    ```

**Process**
1. small(224x224) /  10 epochs / clean dataset(1.6M)
2. medium(512x512) /  30-40 epochs / noisy dataset(3.2M)
3. large(768x768) / a few epochs(finetune) / noisy dataset(3.2M)

---
# sub-center ArcFace - ECCV 2020

[](https://ibug.doc.ic.ac.uk/media/uploads/documents/eccv_1445.pdf)

![image](https://user-images.githubusercontent.com/69357689/193796196-b371649c-6d0d-45ee-ad9d-34dce53bf8a0.png)

- ArcFace
    
    [](https://arxiv.org/pdf/1801.07698.pdf)
    
![image](https://user-images.githubusercontent.com/69357689/193796229-b5936a84-f2f1-4b01-a2e0-c644d2c3f2ff.png)
    
클래스간 각도를 통해 차이를 주어 서로 다른 클래스간에는 더 큰 격차를 만드는 방법

각도 기반으로 margin을 주어 서로 다른 두 클래스 사이에 경계를 분명히 한다.
    
![image](https://user-images.githubusercontent.com/69357689/193796249-fac2a24a-6268-4429-b47a-6e79150ceb13.png)
    
- sub-center ArcFace = ArcFace + introduce sub-class to relax the intra-class constraint →improve the robustness to label noise
- design K sub-centers for each class
and the training sample only needs to be close to any of the K positive subcenters instead of the only one positive center.
- The proposed sub-center ArcFace encourages one **dominant** sub-class that contains the majority of clean faces and non-dominant sub-classes that include hard or noisy faces
- After the model achieves enough discriminative power, we directly drop
non-dominant sub-centers and high-confident noisy samples, which helps recapture intra-compactness, decrease the influence from noise, and achieve comparable performance compared to ArcFace trained on the manually cleaned dataset.

## Training(1st solution)

1. train only image transformer on 224x224 size
2. Exchange original patch embedding module with block 0,1,2 from Efficientnet encoder
3. freeze image transformer and sub-arcface head and train for 1 epoch on 448x448, to let the effnet encoder “adjust”
4. unfreeze image transformer and head and train for 30-40 epochs
5. add block 3 from effnet encoder and finetune a few epochs on 896x896

# TODO
1. Stable diffusion + KELIP text encoder
2. Stable diffusion training for korean text-image pair 
3. Fashion diffusion model demo w/ gradio
